<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

## Lab: Cleaning Rock Song Data

_Authors: Dave Yerrington (SF)_

---


In [1]:
import pandas as pd
import numpy as np 
import seaborn as sns

%matplotlib inline

### 1. Load `rock.csv` and do an initial examination of its data columns.

In [2]:
rockfile = "./datasets/rock.csv"

In [3]:
# Load the data.
rock = pd.read_csv(rockfile)

In [4]:
# Look at the information regarding its columns.
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
Song Clean      2230 non-null object
ARTIST CLEAN    2230 non-null object
Release Year    1653 non-null object
COMBINED        2230 non-null object
First?          2230 non-null int64
Year?           2230 non-null int64
PlayCount       2230 non-null int64
F*G             2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


### 2.  Clean up the column names.

Let's clean up the column names. There are two ways we can accomplish this:

#### 2.A Change the column names when you import the data using `pd.read_csv()`.

Notice that, when passing `names=[..A LIST OF STRING..]` with a number of columns that matches the number of strings in the passed list, you replace the column names.

NOTE: When you create custom column names, the first row of the `.csv` already represents a header. It is important to tell `pandas` to skip that row. The `skiprows=1` keyword argument to `read_csv()` will tell `pandas` to skip the first row.

In [5]:
# Change the column names when loading the '.csv':
names = ['song', 'artist', 'release', 'combined', 'first', 'year', 'playcount', 'fg']
rock = pd.read_csv(rockfile, skiprows=1, names=names)
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
song         2230 non-null object
artist       2230 non-null object
release      1653 non-null object
combined     2230 non-null object
first        2230 non-null int64
year         2230 non-null int64
playcount    2230 non-null int64
fg           2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


#### 2.B Change column names using the `.rename()` function.

The `.rename()` function takes an argument, `columns=name_dict`, in which `name_dict` is a dictionary containing the original column names as keys and the new column names as values.

In [6]:
# Change the column names using the `.rename()` function.

#### 2.C Reassigning the `.columns` attribute of a DataFrame.

You can also just reassign the `.columns` attribute to a list of strings containing the new column names. 

The only caveat with reassigning `.columns` is that you have to reassign all of the column names at once. You can't partially replace a value by working on `.columns` directly. You have to reassign `.columns` with a list of equal length. 

In [7]:
# Replace the column names by reassigning the `.columns` attribute.
rock.columns = names
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
song         2230 non-null object
artist       2230 non-null object
release      1653 non-null object
combined     2230 non-null object
first        2230 non-null int64
year         2230 non-null int64
playcount    2230 non-null int64
fg           2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


### 3. Subsetting data where null values exist.

We have mixed `str` and `NaN` values in the `release` column. `NaN` stands for "not a number" and is the way `pandas` handles "nulls" or nonexistent data. We can use the `.isnull()` method of a Series to find null values.

Print the header of the data subset to where the `release` column is null values.

In [8]:
# Show records where df['release'] is null

rock[rock['release'].isnull()]

song                    artist release  \
1                           Fantasy Girl               .38 Special     NaN   
10                 Baby, Please Don't Go                     AC/DC     NaN   
13                CAN'T STOP ROCK'N'ROLL                     AC/DC     NaN   
16                      Girls Got Rhythm                     AC/DC     NaN   
24                       Let's Get It Up                     AC/DC     NaN   
25                             Live Wire                     AC/DC     NaN   
26                            Moneytalks                     AC/DC     NaN   
29                       Shoot To Thrill                     AC/DC     NaN   
31                              Sin City                     AC/DC     NaN   
35        What Do You Do For Money Honey                     AC/DC     NaN   
42                 Baby, Please Don't Go                 Aerosmith     NaN   
45                         Come Together                 Aerosmith     NaN   
53                      Kings and Queens                 Aerosmith     NaN   
60                        One Way Street                 Aerosmith     NaN   
66                        The Other Side                 Aerosmith     NaN   
71                     You See Me Crying                 Aerosmith     NaN   
75                               Fantasy                 Aldo Nova     NaN   
76                           BE MY LOVER              Alice Cooper     NaN   
80                         House Of Fire              Alice Cooper     NaN   
100                          One Way Out      Allman Brothers Band     NaN   
102                           Southbound      Allman Brothers Band     NaN   
104                        Whippin' Post      Allman Brothers Band     NaN   
110                       Immigrant Song                Ann Wilson     NaN   
111                               Roller                April Wine     NaN   
113              Coming into Los Angeles              Arlo Guthrie     NaN   
114                              Wayside             Artful Dodger     NaN   
118                               Spooky    Atlanta Rhythm Section     NaN   
120                Blondes In Black Cars                 Autograph     NaN   
122   Rock And Roll Party In The Streets                       Axe     NaN   
124             Roll On Down The Highway  Bachman-Turner Overdrive     NaN   
...                                  ...                       ...     ...   
2149              Take Your Whiskey Home                 Van Halen     NaN   
2150                           Top Jimmy                 Van Halen     NaN   
2151                    Top of the World                 Van Halen     NaN   
2157                    And It Stoned Me              Van Morrison     NaN   
2158                          BLUE MONEY              Van Morrison     NaN   
2162                              Gloria              Van Morrison     NaN   
2164                  Jackie Wilson Said              Van Morrison     NaN   
2168              Edge of a Broken Heart                     Vixen     NaN   
2174                   Uncle Tom's Cabin                   Warrant     NaN   
2178              Street Corner Serenade                Wet Willie     NaN   
2179                                Wait                White Lion     NaN   
2184                       Slow An' Easy                Whitesnake     NaN   
2185                  Still of the Night                Whitesnake     NaN   
2186                        Who Are You?                   Why Not     NaN   
2188                         Night Moves                  Wildlife     NaN   
2189                   We've Got Tonight                  Wildlife     NaN   
2190                    Can't Get Enough                    Winger     NaN   
2191             Headed For A Heartbreak                    Winger     NaN   
2192                           Seventeen                    Winger     NaN   
2194                    Summertime Girls                  

### 4. Update slices of your DataFrame based on mask selection/slices.

In many scenarios, we want to upate values in our DataFrame according to criteria. Let's say we wanted to set all of the null values in `release` to 0.

With newer versions of `pandas`, in order to manipulate data in the original DataFrame, we have to use `.loc` while performing reassignment using a mask and an index.

For example, the following won't always work:
```python
df[row_mask]['column_name'] = new_value
```

The best way to accomplish the same task is:
```python
df.loc[row_mask, 'column_name'] = new_value
```

For multiple column assignment, you would use:
```python
df.loc[row_mask, ['col_1', 'col_2', 'col_3']] = new_value
```

#### 4.A Let's try it out. Make all of the null values in `release` 0.

In [9]:
# Replace release nulls with 0
row_mask = rock['release'].isnull()
rock.loc[row_mask, 'release']= np.nan

#### 4.B Verify that `release` contains no null values.

In [10]:
# A:
rock.isnull().sum()

song           0
artist         0
release      577
combined       0
first          0
year           0
playcount      0
fg             0
dtype: int64

### 5. Ensure that the data types of the columns make sense. 

Verifying column data types is a critical part of data munging. If columns have the wrong data type, then there is usually corrupted or incorrect data in some of the observations.

#### 5.A Look at the data types for the columns. Are any incorrect given what the data represents?

In [11]:
# A:
names = ['song', 'artist', 'release', 'combined', 'first', 'year', 'playcount', 'fg']
rock = pd.read_csv(rockfile, skiprows=1, names=names)
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
song         2230 non-null object
artist       2230 non-null object
release      1653 non-null object
combined     2230 non-null object
first        2230 non-null int64
year         2230 non-null int64
playcount    2230 non-null int64
fg           2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


### 6. Investigate and clean up the `release` column.

The `release` column is a string data type when it should be an integer.

#### 6.A Figure out what value(s) are causing the `release` column to be encoded as a string instead of an integer.

In [12]:
# A:

#### 6.B Look at the rows in which there is incorrect data in the `release` column.

In [13]:
# A:
rock.release.unique()

array(['1982', nan, '1981', '1980', '1975', '2000', '2002', '1992',
       '1985', '1993', '1976', '1995', '1979', '1984', '1977', '1990',
       '1986', '1974', '2014', '1987', '1973', '2001', '1989', '1997',
       '1971', '1972', '1994', '1970', '1966', '1965', '1983', '1955',
       '1978', '1969', '1999', '1968', '1988', '1962', '2007', '1967',
       '1958', '1071', '1996', '1991', '2005', '2011', '2004', '2012',
       '2003', '1998', '2008', '1964', '2013', '2006', 'SONGFACTS.COM',
       '1963', '1961'], dtype=object)

#### 6.C. Clean up the data. Normally we may replace the offending data with null np.nan values, however we previously converted all of the nan values in the release column to zeros so we might as well continue with the same practice. Replacing with 0 (or nan) will allow us to convert the column to numeric.

In [14]:
# A:
row_incorrect = rock['release'] == 'SONGFACTS.COM'
rock.loc[row_incorrect, 'release'] = np.nan
rock['release'] = rock['release'].map(lambda x: float(x))

### 7. Get summary statistics for the `release` column using the `.describe()` function.

Now that the `release` column is finally a numeric data type, we can apply the `.describe()` function.  

#### 7.A Print out the summary stats for the `release` column. What is the earliest and latest release date?

In [15]:
# A:
rock['release'].describe()

count    1652.000000
mean     1978.019976
std        24.191247
min      1071.000000
25%      1971.000000
50%      1977.000000
75%      1984.000000
max      2014.000000
Name: release, dtype: float64

#### 7.B Based on the summary statistics, is there anything else wrong with the `release` column? 

In [16]:
# A:
rock.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
song         2230 non-null object
artist       2230 non-null object
release      1652 non-null float64
combined     2230 non-null object
first        2230 non-null int64
year         2230 non-null int64
playcount    2230 non-null int64
fg           2230 non-null int64
dtypes: float64(1), int64(4), object(3)
memory usage: 139.5+ KB


_Looking at the DataFrame that contains the year 1071, we can see that the year was probably corrupted and should be replaced with something else if possible._

### 8. Make changes and investigate using custom functions with `.apply()`.

Let's say we want to traverse every single row in our data set and apply a function to that row.

#### 8.A Write a function that will take a row of a DataFrame and print out the song, artist, and whether or not the release date is < 1970.


In [17]:
# A:


def before1970 (row):
    print(row['song'], row['artist'], row['release'], '<1970?', row['release'] < 1970)
    
    
# def release_inspector(row):
#     print('--------------------------------')
#     print(row['song'], row['artist'], row['release'], '< 1970?', row['release'] < 1970)

#### 8.B Using the `.apply()` function, apply the function you wrote to the first four rows of the DataFrame.

You will need to tell the `apply` function to operate row by row. Setting the keyword argument as `axis=1` indicates that the function should be applied to each row individually.

In [18]:
# A:
names = ['song', 'artist', 'release', 'combined', 'first', 'year', 'playcount', 'fg']
rock = pd.read_csv(rockfile, names=names, skiprows=1)

# column_names = ['song', 'artist', 'release', 'song_artist', 'first', 'year', 'playcount', 'fg']
# df = pd.read_csv(rockfile, names=column_names, skiprows=1)

# rock.head(4).apply(before1970, axis=1)
rock.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
song         2230 non-null object
artist       2230 non-null object
release      1653 non-null object
combined     2230 non-null object
first        2230 non-null int64
year         2230 non-null int64
playcount    2230 non-null int64
fg           2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.5+ KB


You'll notice that there will be a final output Series of `None` values. The `.apply()` function, if a return value is not specified, will return a Series of `None` values (similar to how the default return for Python functions is `None` when a return statement is not specified).

### 9. Write a function that converts cells in a DataFrame to float and otherwise replaces them with `np.nan`.

If applied to our data, it would keep only the numeric information and otherwise input null values.

Recall that the try-except syntax in Python is a great way to try something and take another action if the initial step fails:

```python
try:
    Perform some action.
except:
   Perform some other action if the first failed with an error.
```

#### 9.A Write the function that takes a column and converts all of its values to float if possible and `np.nan` otherwise. The return value should be the converted Series.

In [19]:
# A:
def floatcolumn(value):
    try:
        return float(value)
    except:
        return np.nan
def columnconvert(column):
    column = column.map(floatcolumn)
    return column

#### 9.B Try your function out on the rock song data and ensure the output is what you expected.


In [20]:
# A
rock.apply(columnconvert)

song  artist  release  combined  first  year  playcount     fg
0      NaN     NaN   1982.0       NaN    1.0   1.0       82.0   82.0
1      NaN     NaN      NaN       NaN    1.0   0.0        3.0    0.0
2      NaN     NaN   1981.0       NaN    1.0   1.0       85.0   85.0
3      NaN     NaN   1980.0       NaN    1.0   1.0       18.0   18.0
4      NaN     NaN   1975.0       NaN    1.0   1.0        1.0    1.0
5      NaN     NaN   2000.0       NaN    1.0   1.0       13.0   13.0
6      NaN     NaN   2000.0       NaN    1.0   1.0        1.0    1.0
7      NaN     NaN   2002.0       NaN    1.0   1.0        6.0    6.0
8      NaN     NaN   1992.0       NaN    1.0   1.0        3.0    3.0
9      NaN     NaN   1985.0       NaN    1.0   1.0        1.0    1.0
10     NaN     NaN      NaN       NaN    1.0   0.0        1.0    0.0
11     NaN     NaN   1980.0       NaN    1.0   1.0       97.0   97.0
12     NaN     NaN   1993.0       NaN    1.0   1.0        6.0    6.0
13     NaN     NaN      NaN       NaN    1.0   0.0        5.0    0.0
14     NaN     NaN   1976.0       NaN    1.0   1.0       85.0   85.0
15     NaN     NaN   1981.0       NaN    1.0   1.0       46.0   46.0
16     NaN     NaN      NaN       NaN    1.0   0.0       24.0    0.0
17     NaN     NaN   1995.0       NaN    1.0   1.0        1.0    1.0
18     NaN     NaN   1980.0       NaN    1.0   1.0       39.0   39.0
19     NaN     NaN   1980.0       NaN    1.0   1.0       74.0   74.0
20     NaN     NaN   1979.0       NaN    1.0   1.0       92.0   92.0
21     NaN     NaN   1975.0       NaN    1.0   1.0       39.0   39.0
22     NaN     NaN   1984.0       NaN    1.0   1.0        1.0    1.0
23     NaN     NaN   1977.0       NaN    1.0   1.0        3.0    3.0
24     NaN     NaN      NaN       NaN    1.0   0.0        4.0    0.0
25     NaN     NaN      NaN       NaN    1.0   0.0        2.0    0.0
26     NaN     NaN      NaN       NaN    1.0   0.0       20.0    0.0
27     NaN     NaN   1979.0       NaN    1.0   1.0        1.0    1.0
28     NaN     NaN   1980.0       NaN    1.0   1.0       21.0   21.0
29     NaN     NaN      NaN       NaN    1.0   0.0       45.0    0.0
...    ...     ...      ...       ...    ...   ...        ...    ...
2200   NaN     NaN   1983.0       NaN    1.0   1.0       81.0   81.0
2201   NaN     NaN   1971.0       NaN    1.0   1.0       44.0   44.0
2202   NaN     NaN   1971.0       NaN    1.0   1.0        1.0    1.0
2203   NaN     NaN      NaN       NaN    1.0   0.0        5.0    0.0
2204   NaN     NaN   1971.0       NaN    1.0   1.0        1.0    1.0
2205   NaN     NaN      NaN       NaN    1.0   0.0        2.0    0.0
2206   NaN     NaN      NaN       NaN    1.0   0.0        1.0    0.0
2207   NaN     NaN   1968.0       NaN    1.0   1.0       14.0   14.0
2208   NaN     NaN      NaN       NaN    1.0   0.0        1.0    0.0
2209   NaN     NaN   1976.0       NaN    1.0   1.0        1.0    1.0
2210   NaN     NaN   1973.0       NaN    1.0   1.0        8.0    8.0
2211   NaN     NaN   1980.0       NaN    1.0   1.0       34.0   34.0
2212   NaN     NaN   1983.0       NaN    1.0   1.0       86.0   86.0
2213   NaN     NaN   1983.0       NaN    1.0   1.0       29.0   29.0
2214   NaN     NaN   1975.0       NaN    1.0   1.0        2.0    2.0
2215   NaN     NaN      NaN       NaN    1.0   0.0       16.0    0.0
2216   NaN     NaN      NaN       NaN    1.0   0.0       10.0    0.0
2217   NaN     NaN   1973.0       NaN    1.0   1.0        6.0    6.0
2218   NaN     NaN      NaN       NaN    1.0   0.0        2.0    0.0
2219   NaN     NaN   1973.0       NaN    1.0   1.0      111.0  111.0
2220   NaN     NaN   1983.0       NaN    1.0   1.0      121.0  121.0
2221   NaN     NaN      NaN       NaN    1.0   0.0        1.0    0.0
2222   NaN     NaN      NaN       NaN    1.0   0.0       14.0    0.0
2223   NaN     NaN   1981.0       NaN    1.0   1.0        5.0    5.0
2224   NaN     NaN   1983.0       NaN    1.0   1.0      120.0  120.0
2225   NaN     NaN      NaN       NaN    1.0   0.0        1.0    0.0
2226

#### 9.C Describe the new float-only DataFrame.

In [21]:
# A:
rock2 = rock.apply(columnconvert)

In [22]:
rock2.describe()

song  artist      release  combined   first         year  \
count     2.000000     0.0  1652.000000       0.0  2230.0  2230.000000   
mean   1012.000000     NaN  1978.019976       NaN     1.0     0.741256   
std    1367.544515     NaN    24.191247       NaN     0.0     0.438043   
min      45.000000     NaN  1071.000000       NaN     1.0     0.000000   
25%     528.500000     NaN  1971.000000       NaN     1.0     0.000000   
50%    1012.000000     NaN  1977.000000       NaN     1.0     1.000000   
75%    1495.500000     NaN  1984.000000       NaN     1.0     1.000000   
max    1979.000000     NaN  2014.000000       NaN     1.0     1.000000   

         playcount           fg  
count  2230.000000  2230.000000  
mean     16.872646    15.048430  
std      25.302972    25.288366  
min       0.000000     0.000000  
25%       1.000000     0.000000  
50%       4.000000     3.000000  
75%      21.000000    18.000000  
max     142.000000   142.000000